___

<a href='http://www.pieriandata.com'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Авторские права принадлежат Pierian Data Inc.</em></center>
<center><em>Для дополнительной информации посетите наш сайт <a href='http://www.pieriandata.com'>www.pieriandata.com</a></em></center>

# Датафреймы - DataFrames

На протяжении всего курсы мы будем очень много работать с объектами DataFrame. Они являются естественным продолжением Pandas Series. Определение DataFrame следующее:

**Pandas DataFrame состоит из нескольких объектов Pandas Series с общими значениями индекса.**  
  
  
По сути, если Series - это **одномерный** проиндексированный массив, то DataFrame - **многомерный** проиндексированный массив, в котором каждый столбец является массивом Series.

Давайте рассмотрим на примерах.

## Imports

In [424]:
import numpy as np
import pandas as pd

### Создаём DataFrame на основе объектов Python

In [425]:
# help(pd.DataFrame)

In [426]:
# Убедитесь, что команда seed находится в той же ячейке, что и команда random
# https://stackoverflow.com/questions/21494489/what-does-numpy-random-seed0-do
np.random.seed(101)
mydata = np.random.randint(0,101,(4,3))

In [427]:
mydata

array([[95, 11, 81],
       [70, 63, 87],
       [75,  9, 77],
       [40,  4, 63]])

In [428]:
myindex = ['CA','NY','AZ','TX']

In [429]:
mycolumns = ['Jan','Feb','Mar']

In [430]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0  95  11  81
1  70  63  87
2  75   9  77
3  40   4  63

In [431]:
df = pd.DataFrame(data=mydata,index=myindex)
df

0   1   2
CA  95  11  81
NY  70  63  87
AZ  75   9  77
TX  40   4  63

In [432]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df 

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [433]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int32
 1   Feb     4 non-null      int32
 2   Mar     4 non-null      int32
dtypes: int32(3)
memory usage: 80.0+ bytes



# Загружаем файл .csv file в DataFrame

----

## Замечание: Позже в этом курсе мы рассмотрим много различных типов файлов - .html, .csv, .xlsx и другие! Но сейчас мы ограничимся только файлом .csv.

----

## CSV
Файлы CSV (Comma Separated Values) содержат отдельные значения данных, отделённые друг от друга с помощью знаков-разделителей.<br>
Если Вы НЕ используете виртуальную среду, предлагаемую в этом курсе, то Вам может потребоваться установка библиотек <tt>xlrd</tt> и <tt>openpyxl</tt>.<br>
В Вашем терминале или командной строке выполните следующие команды:

    conda install xlrd
    conda install openpyxl

(если Вы не используете Anaconda Distribution, то вместо conda install напишите pip install)

После этого перезапустите Jupyter Notebook.

### Путь к файлу

Когда нужно обратиться к файлу в pandas, то есть два варианта:

1. Если Ваш файл .csv находится **в той же самой** папке, что и файл .py или блокнот .ipynb, то можно просто указать название этого файла .csv. Например:
    
        df = pd.read_csv('some_file.csv')
        
2. Либо второй вариант - если файл .csv находится в другой папке, то можно указать полный путь к файлу. Обратите внимание, что путь должен быть на 100% правильным, без единой опечатки. Например:

        df = pd.read_csv("C:\\Users\\myself\\files\\some_file.csv")

#### Отобразить текущую папку можно с помощью команды pwd

In [434]:
pwd

'C:\\work\\jupiter_ML\\ML_course\\UNZIP_ME_FOR_NOTEBOOKS_ML_RUS_V9\\03-Pandas'

#### Вывести список файлов в текущей папке - команда ls

In [435]:
ls

 ’®¬ ў гбва®©б⢥ C Ё¬ҐҐв ¬ҐвЄг Windows-SSD
 ‘ҐаЁ©­л© ­®¬Ґа ⮬ : 4E76-F254

 ‘®¤Ґа¦Ё¬®Ґ Ї ЇЄЁ C:\work\jupiter_ML\ML_course\UNZIP_ME_FOR_NOTEBOOKS_ML_RUS_V9\03-Pandas

10.09.2024  23:46    <DIR>          .
08.09.2024  12:01    <DIR>          ..
10.09.2024  22:45    <DIR>          .ipynb_checkpoints
08.09.2024  12:01    <DIR>          __pycache__
10.09.2024  21:52         1я301я173 00-Series.ipynb
10.09.2024  23:46           332я354 01-DataFrames.ipynb
08.09.2024  12:01           195я623 02-Conditional-Filtering.ipynb
08.09.2024  12:01           196я925 03-Useful-Methods.ipynb
08.09.2024  12:01            65я702 04-Missing-Data.ipynb
08.09.2024  12:01           247я978 05-Groupby-Operations-and-MultiIndex.ipynb
08.09.2024  12:01            66я225 06-Combining-DataFrames.ipynb
08.09.2024  12:01            34я195 07-Text-Methods.ipynb
08.09.2024  12:01            96я529 08-Time-Methods.ipynb
08.09.2024  12:01            70я505 09-Inputs-and-Outputs.ipynb
08.09.2024  12:01           103я40

In [436]:
df = pd.read_csv('tips.csv')

In [437]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
Подробности об этом наборе данных (если Вам интересно)

* Описание
    * Один официант записывал информацию о чаевых, которые он получал с течении нескольких месяцев, работая в ресторане. 

* Формат
    * 244 строк и 7 переменных

* Для каждого факта получения чаевых этот официант записал следующую информацию:
    * Сумма чаевых в долларах (tip) 
    * Сумма счёта в долларах (total_bill)
    * Пол человека, оплатившего счёт (sex)
    * Были ли в компании курящие (smoker)
    * День недели (day)
    * Время дня (time)
    * Количество человек в компании (size)

Всего он сделал 244 записи. Эти данные были использованы для исследований по применению статистики для бизнес-задач (Bryant & Smith 1995).

* Литература
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Замечание: Мы добавили несколько колонок с вымышленными данными - имя (Name), номер банковской карты (CC Number) и идентификатор платежа (Payment ID).

----

# DataFrames

## Как получить базовую информацию об объекте DataFrame

In [438]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [439]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [440]:
df.head(3)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458

In [441]:
df.tail(3)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [442]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [443]:
len(df)

244

In [444]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [445]:
df.describe().transpose()

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

## Индексирование

Давайте узнаем, как можно извлекать данные из DataFrame.

### КОЛОНКИ

Начнём с того, как получить данные на основе колонок

In [446]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Берём одну колонку

In [447]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [448]:
type(df['total_bill'])

pandas.core.series.Series

#### Берём несколько колонок

In [449]:
# Обратите внимание, что здесь мы передаём названия колонок в виде списка (list). 
# Поэтому здесь два набора квадратных скобок.
df[['total_bill','tip']]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Создание новых колонок

In [450]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

In [451]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [452]:
df['price_per_person'] = df['total_bill'] / df['size']

In [453]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.66    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.50    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.31    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.61  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [454]:
help(np.round)

Help on _ArrayFunctionDispatcher in module numpy:

round(a, decimals=0, out=None)
    Evenly round to the given number of decimals.

    Parameters
    ----------
    a : array_like
        Input data.
    decimals : int, optional
        Number of decimal places to round to (default: 0).  If
        decimals is negative, it specifies the number of positions to
        the left of the decimal point.
    out : ndarray, optional
        Alternative output array in which to place the result. It must have
        the same shape as the expected output, but the type of the output
        values will be cast if necessary. See :ref:`ufuncs-output-type` for more
        details.

    Returns
    -------
    rounded_array : ndarray
        An array of the same type as `a`, containing the rounded values.
        Unless `out` was specified, a new array is created.  A reference to
        the result is returned.

        The real and imaginary parts of complex numbers are rounded
        separately

#### Изменение данных в существующих колонках

In [455]:
# Поскольку pandas основан на numpy, мы можем очень удобно применять универсальные функции numpy!
df['price_per_person'] = np.round(df['price_per_person'],2)

In [456]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

#### Удаление колонок

In [457]:
# df.drop('tip_percentage',axis=1)

In [458]:
df = df.drop("tip_percentage",axis=1)

In [459]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

# Основы работы с индексами

Прежде чем перейти в аналогичным действиям не с колонками, а со строками, давайте сначала разберёмся, как работает индексация в pandas DataFrame.

In [460]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [461]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [462]:
# Обратите внимание - здесь мы не меняем df, а только отображаем результат метода set_index
df.set_index('Payment ID')

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221  
...                      ...                 ...               ...  
Sat2657                 9.68       Michael Avila  5296068606052842  
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  

[244 rows x 10 columns]

In [463]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [464]:
# А теперь поменяем сам df - сохраним в нём результат работы метода set_index
df = df.set_index('Payment ID')

In [465]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [466]:
df = df.reset_index()

In [467]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

### СТРОКИ

Теперь давайте применим те же идеи для строк.

In [468]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

In [469]:
df = df.set_index('Payment ID')

In [470]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

#### Берём одну строку

In [471]:
# На основе номера строки - числа Integer
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [472]:
# На основе названия (на основе значения колонки-индекса)
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

#### Берём несколько строк

In [473]:
df.iloc[0:4]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994

In [474]:
df.loc[['Sun2959','Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun5260                11.84    Nathaniel Harris  4676137647685994

#### Удаляем строки

Поскольку обычно наборы данных содержат очень много строк, мы почти никогда не удаляем строки, указывая номера строк. Вместо этого мы обычно удаляем строки на основе каких-то условий, например если какие-то данные отсутствуют в колонках. В следующей лекции мы разберём это более подробно.

In [475]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [476]:
df.drop('Sun2959',axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun4608                 3.45    Douglas Tucker  4478071379779230  
Sun4458                 7.00    Travis Walters  6011812112971322  
Sun5260                11.84  Nathaniel Harris  4676137647685994  
Sun2251                 6.15      Tonya Carter  4832732618637221  
Sun9679                 6.32        Erik Smith   213140353657882

In [477]:
# Приведёт к ошибке, если у Вас есть именованный индекс!
# df.drop(0,axis=0).head()

#### Вставка новой строки

Такая вставка одной строки встречается очень редко. Чаще используется pd.concat() для вставки сразу нескольких строк.  
Ещё можно применить метод .append() для списка объектов pd.Series(), но это тоже скорее редкие случаи.

In [478]:
one_row = df.iloc[0]

In [479]:
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [480]:
type(one_row)

pandas.core.series.Series

In [481]:
df.tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  
Payment ID                                                       
Sat2657                 9.68    Michael Avila  5296068606052842  
Sat1766                13.59   Monica Sanders  3506806155565404  
Sat3880                11.34       Keith Wong  6011891618747196  
Sat17                   8.91     Dennis Dixon     4375220550950  
Thur672                 9.39  Michelle Hardin  3511451626698139

In [482]:
pd.concat([df, pd.DataFrame([one_row])], axis=0).tail()

# Вместо df.append в будущих версиях Pandas следует использовать pd.concat().
# Вот пример кода - здесь one_row это объект Series:
# pd.concat([df,pd.DataFrame([one_row])],axis=0)

total_bill   tip     sex smoker   day    time  size  \
Sat1766       27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880       22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17         17.82  1.75    Male     No   Sat  Dinner     2   
Thur672       18.78  3.00  Female     No  Thur  Dinner     2   
Sun2959       16.99  1.01  Female     No   Sun  Dinner     2   

         price_per_person          Payer Name         CC Number  
Sat1766             13.59      Monica Sanders  3506806155565404  
Sat3880             11.34          Keith Wong  6011891618747196  
Sat17                8.91        Dennis Dixon     4375220550950  
Thur672              9.39     Michelle Hardin  3511451626698139  
Sun2959              8.49  Christy Cunningham  3560325168603410

In [483]:


# Вместо df.append в будущих версиях Pandas следует использовать pd.concat().
# Вот пример кода - здесь one_row это объект Series:
# pd.concat([df,pd.DataFrame([one_row])],axis=0)

In [3]:
print('Отличная Работа')

In [485]:
import numpy as np

In [486]:
import pandas as pd

In [487]:
np.random.seed(101)
mydata = np.random.randint(0, 101, (4, 3))
mydata

array([[95, 11, 81],
       [70, 63, 87],
       [75,  9, 77],
       [40,  4, 63]])

In [488]:
myindex = ['CA', 'NY', 'AZ', 'TX']

In [489]:
mycolumns = ['Jan', 'Feb', 'Mar']

In [490]:
df_0 = pd.DataFrame(mydata)

In [491]:
df_0

0   1   2
0  95  11  81
1  70  63  87
2  75   9  77
3  40   4  63

In [492]:
df = pd.DataFrame(mydata, myindex, mycolumns)

In [493]:
df

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [494]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int32
 1   Feb     4 non-null      int32
 2   Mar     4 non-null      int32
dtypes: int32(3)
memory usage: 80.0+ bytes


In [495]:
pwd # где находится файл

'C:\\work\\jupiter_ML\\ML_course\\UNZIP_ME_FOR_NOTEBOOKS_ML_RUS_V9\\03-Pandas'

In [496]:
ls

 ’®¬ ў гбва®©б⢥ C Ё¬ҐҐв ¬ҐвЄг Windows-SSD
 ‘ҐаЁ©­л© ­®¬Ґа ⮬ : 4E76-F254

 ‘®¤Ґа¦Ё¬®Ґ Ї ЇЄЁ C:\work\jupiter_ML\ML_course\UNZIP_ME_FOR_NOTEBOOKS_ML_RUS_V9\03-Pandas

10.09.2024  23:46    <DIR>          .
08.09.2024  12:01    <DIR>          ..
10.09.2024  22:45    <DIR>          .ipynb_checkpoints
08.09.2024  12:01    <DIR>          __pycache__
10.09.2024  21:52         1я301я173 00-Series.ipynb
10.09.2024  23:46           332я354 01-DataFrames.ipynb
08.09.2024  12:01           195я623 02-Conditional-Filtering.ipynb
08.09.2024  12:01           196я925 03-Useful-Methods.ipynb
08.09.2024  12:01            65я702 04-Missing-Data.ipynb
08.09.2024  12:01           247я978 05-Groupby-Operations-and-MultiIndex.ipynb
08.09.2024  12:01            66я225 06-Combining-DataFrames.ipynb
08.09.2024  12:01            34я195 07-Text-Methods.ipynb
08.09.2024  12:01            96я529 08-Time-Methods.ipynb
08.09.2024  12:01            70я505 09-Inputs-and-Outputs.ipynb
08.09.2024  12:01           103я40

In [497]:
df_tips = pd.read_csv('tips.csv') 

In [498]:
df

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [499]:
df_tips

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

In [500]:
df_tips.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [501]:
df_tips.index

RangeIndex(start=0, stop=244, step=1)

In [502]:
df_tips.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [503]:
df_tips.tail()

total_bill   tip     sex smoker   day    time  size  price_per_person  \
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
239    Michael Avila  5296068606052842    Sat2657  
240   Monica Sanders  3506806155565404    Sat1766  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [504]:
df_tips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [505]:
df.describe()

Jan        Feb        Mar
count   4.000000   4.000000   4.000000
mean   70.000000  21.750000  77.000000
std    22.730303  27.657127  10.198039
min    40.000000   4.000000  63.000000
25%    62.500000   7.750000  73.500000
50%    72.500000  10.000000  79.000000
75%    80.000000  24.000000  82.500000
max    95.000000  63.000000  87.000000

In [506]:
df_tips.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

In [507]:
df_tips.T

0                 1                 2    \
total_bill                     16.99             10.34             21.01   
tip                             1.01              1.66               3.5   
sex                           Female              Male              Male   
smoker                            No                No                No   
day                              Sun               Sun               Sun   
time                          Dinner            Dinner            Dinner   
size                               2                 3                 3   
price_per_person                8.49              3.45               7.0   
Payer Name        Christy Cunningham    Douglas Tucker    Travis Walters   
CC Number           3560325168603410  4478071379779230  6011812112971322   
Payment ID                   Sun2959           Sun4608           Sun4458   

                               3                 4                5    \
total_bill                   23.68             24.59            25.29   
tip                           3.31              3.61             4.71   
sex                           Male            Female             Male   
smoker                          No                No               No   
day                            Sun               Sun              Sun   
time                        Dinner            Dinner           Dinner   
size                             2                 4                4   
price_per_person             11.84              6.15             6.32   
Payer Name        Nathaniel Harris      Tonya Carter       Erik Smith   
CC Number         4676137647685994  4832732618637221  213140353657882   
Payment ID                 Sun5260           Sun2251          Sun9679   

                                 6                 7                 8    \
total_bill                      8.77             26.88             15.04   
tip                              2.0              3.12              1.96   
sex                             Male              Male              Male   
smoker                            No                No                No   
day                              Sun               Sun               Sun   
time                          Dinner            Dinner            Dinner   
size                               2                 4                 2   
price_per_person                4.38              6.72              7.52   
Payer Name        Kristopher Johnson       Robert Buck   Joseph Mcdonald   
CC Number           2223727524230344  3514785077705092  3522866365840377   
Payment ID                   Sun5985           Sun8157           Sun6820   

                               9    ...               234               235  \
total_bill                   14.78  ...             15.53             10.07   
tip                           3.23  ...               3.0              1.25   
sex                           Male  ...              Male              Male   
smoker                          No  ...               Yes                No   
day                            Sun  ...               Sat               Sat   
time                        Dinner  ...            Dinner            Dinner   
size                             2  ...                 2                 2   
price_per_person              7.39  ...              7.76              5.04   
Payer Name           Jerome Abbott  ...     Tracy Douglas     Sean Gonzalez   
CC Number         3532124519049786  ...  4097938155941930  3534021246117605   
Payment ID                 Sun3775  ...           Sat7220           Sat4615   

                               236               237             238  \
total_bill                    12.6             32.83           35.83   
tip                            1.0              1.17            4.67   
sex                           Male              Male          Female   
smoker                         Yes               Yes              No   
day                  

In [508]:
df_tips

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

In [509]:
df_tips.transpose()

0                 1                 2    \
total_bill                     16.99             10.34             21.01   
tip                             1.01              1.66               3.5   
sex                           Female              Male              Male   
smoker                            No                No                No   
day                              Sun               Sun               Sun   
time                          Dinner            Dinner            Dinner   
size                               2                 3                 3   
price_per_person                8.49              3.45               7.0   
Payer Name        Christy Cunningham    Douglas Tucker    Travis Walters   
CC Number           3560325168603410  4478071379779230  6011812112971322   
Payment ID                   Sun2959           Sun4608           Sun4458   

                               3                 4                5    \
total_bill                   23.68             24.59            25.29   
tip                           3.31              3.61             4.71   
sex                           Male            Female             Male   
smoker                          No                No               No   
day                            Sun               Sun              Sun   
time                        Dinner            Dinner           Dinner   
size                             2                 4                4   
price_per_person             11.84              6.15             6.32   
Payer Name        Nathaniel Harris      Tonya Carter       Erik Smith   
CC Number         4676137647685994  4832732618637221  213140353657882   
Payment ID                 Sun5260           Sun2251          Sun9679   

                                 6                 7                 8    \
total_bill                      8.77             26.88             15.04   
tip                              2.0              3.12              1.96   
sex                             Male              Male              Male   
smoker                            No                No                No   
day                              Sun               Sun               Sun   
time                          Dinner            Dinner            Dinner   
size                               2                 4                 2   
price_per_person                4.38              6.72              7.52   
Payer Name        Kristopher Johnson       Robert Buck   Joseph Mcdonald   
CC Number           2223727524230344  3514785077705092  3522866365840377   
Payment ID                   Sun5985           Sun8157           Sun6820   

                               9    ...               234               235  \
total_bill                   14.78  ...             15.53             10.07   
tip                           3.23  ...               3.0              1.25   
sex                           Male  ...              Male              Male   
smoker                          No  ...               Yes                No   
day                            Sun  ...               Sat               Sat   
time                        Dinner  ...            Dinner            Dinner   
size                             2  ...                 2                 2   
price_per_person              7.39  ...              7.76              5.04   
Payer Name           Jerome Abbott  ...     Tracy Douglas     Sean Gonzalez   
CC Number         3532124519049786  ...  4097938155941930  3534021246117605   
Payment ID                 Sun3775  ...           Sat7220           Sat4615   

                               236               237             238  \
total_bill                    12.6             32.83           35.83   
tip                            1.0              1.17            4.67   
sex                           Male              Male          Female   
smoker                         Yes               Yes              No   
day                  

In [510]:
df_tips.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [511]:
df_tips['total_bill ']

KeyError: 'total_bill '

In [ ]:
type(df_tips['total_bill'])

In [512]:
mycols = ['total_bill', 'tip']

In [513]:
df_tips[mycols]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

In [514]:
df_tips[['total_bill', 'tip']]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

In [515]:
df_tips['tip_percentage'] = 100 * df_tips['tip'] / df_tips['total_bill']

In [516]:
df_tips.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [517]:
df_tips['price_per_person'] = df_tips['price_per_person'] + 1

In [518]:
df_tips

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              9.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              4.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              8.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             12.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              7.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3             10.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             14.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             12.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              9.91   
243       18.78  3.00  Female     No  Thur  Dinner     2             10.39   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608       16.054159  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4          Tonya Carter  4832732618637221    Sun2251       14.680765  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.392697  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17        9.820426  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [519]:
df_tips

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              9.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              4.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              8.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             12.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              7.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3             10.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             14.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             12.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              9.91   
243       18.78  3.00  Female     No  Thur  Dinner     2             10.39   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608       16.054159  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4          Tonya Carter  4832732618637221    Sun2251       14.680765  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.392697  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17        9.820426  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [520]:
np.round(df_tips, 2)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              9.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              4.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              8.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             12.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              7.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3             10.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             14.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             12.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              9.91   
243       18.78  3.00  Female     No  Thur  Dinner     2             10.39   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959            5.94  
1        Douglas Tucker  4478071379779230    Sun4608           16.05  
2        Travis Walters  6011812112971322    Sun4458           16.66  
3      Nathaniel Harris  4676137647685994    Sun5260           13.98  
4          Tonya Carter  4832732618637221    Sun2251           14.68  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657           20.39  
240      Monica Sanders  3506806155565404    Sat1766            7.36  
241          Keith Wong  6011891618747196    Sat3880            8.82  
242        Dennis Dixon     4375220550950      Sat17            9.82  
243     Michelle Hardin  3511451626698139    Thur672           15.97  

[244 rows x 12 columns]

In [521]:
df_tips

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              9.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              4.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              8.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             12.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              7.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3             10.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             14.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             12.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              9.91   
243       18.78  3.00  Female     No  Thur  Dinner     2             10.39   

             Payer Name         CC Number Payment ID  tip_percentage  
0    Christy Cunningham  3560325168603410    Sun2959        5.944673  
1        Douglas Tucker  4478071379779230    Sun4608       16.054159  
2        Travis Walters  6011812112971322    Sun4458       16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4          Tonya Carter  4832732618637221    Sun2251       14.680765  
..                  ...               ...        ...             ...  
239       Michael Avila  5296068606052842    Sat2657       20.392697  
240      Monica Sanders  3506806155565404    Sat1766        7.358352  
241          Keith Wong  6011891618747196    Sat3880        8.822232  
242        Dennis Dixon     4375220550950      Sat17        9.820426  
243     Michelle Hardin  3511451626698139    Thur672       15.974441  

[244 rows x 12 columns]

In [522]:
df_tips = df_tips.drop('tip_percentage', axis=1)

In [523]:
df

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [524]:
df_tips

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              9.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              4.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              8.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             12.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              7.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3             10.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             14.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             12.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              9.91   
243       18.78  3.00  Female     No  Thur  Dinner     2             10.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

In [525]:
df_tips.shape

(244, 11)

In [526]:
df_tips.index

RangeIndex(start=0, stop=244, step=1)

In [531]:
df_tips.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              9.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              4.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              8.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             12.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              7.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [538]:
df_tips = df_tips.set_index('Payment ID')

In [539]:
df_tips

index  total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                               
Sun2959         0       16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608         1       10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458         2       21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260         3       23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251         4       24.59  3.61  Female     No   Sun  Dinner     4   
...           ...         ...   ...     ...    ...   ...     ...   ...   
Sat2657       239       29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766       240       27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880       241       22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17         242       17.82  1.75    Male     No   Sat  Dinner     2   
Thur672       243       18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 9.49  Christy Cunningham  3560325168603410  
Sun4608                 4.45      Douglas Tucker  4478071379779230  
Sun4458                 8.00      Travis Walters  6011812112971322  
Sun5260                12.84    Nathaniel Harris  4676137647685994  
Sun2251                 7.15        Tonya Carter  4832732618637221  
...                      ...                 ...               ...  
Sat2657                10.68       Michael Avila  5296068606052842  
Sat1766                14.59      Monica Sanders  3506806155565404  
Sat3880                12.34          Keith Wong  6011891618747196  
Sat17                   9.91        Dennis Dixon     4375220550950  
Thur672                10.39     Michelle Hardin  3511451626698139  

[244 rows x 11 columns]

In [540]:
df_tips = df_tips.reset_index()

In [541]:
df_tips

Payment ID  index  total_bill   tip     sex smoker   day    time  size  \
0      Sun2959      0       16.99  1.01  Female     No   Sun  Dinner     2   
1      Sun4608      1       10.34  1.66    Male     No   Sun  Dinner     3   
2      Sun4458      2       21.01  3.50    Male     No   Sun  Dinner     3   
3      Sun5260      3       23.68  3.31    Male     No   Sun  Dinner     2   
4      Sun2251      4       24.59  3.61  Female     No   Sun  Dinner     4   
..         ...    ...         ...   ...     ...    ...   ...     ...   ...   
239    Sat2657    239       29.03  5.92    Male     No   Sat  Dinner     3   
240    Sat1766    240       27.18  2.00  Female    Yes   Sat  Dinner     2   
241    Sat3880    241       22.67  2.00    Male    Yes   Sat  Dinner     2   
242      Sat17    242       17.82  1.75    Male     No   Sat  Dinner     2   
243    Thur672    243       18.78  3.00  Female     No  Thur  Dinner     2   

     price_per_person          Payer Name         CC Number  
0                9.49  Christy Cunningham  3560325168603410  
1                4.45      Douglas Tucker  4478071379779230  
2                8.00      Travis Walters  6011812112971322  
3               12.84    Nathaniel Harris  4676137647685994  
4                7.15        Tonya Carter  4832732618637221  
..                ...                 ...               ...  
239             10.68       Michael Avila  5296068606052842  
240             14.59      Monica Sanders  3506806155565404  
241             12.34          Keith Wong  6011891618747196  
242              9.91        Dennis Dixon     4375220550950  
243             10.39     Michelle Hardin  3511451626698139  

[244 rows x 12 columns]

In [542]:
df_tips.head()

Payment ID  index  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959      0       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608      1       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458      2       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260      3       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251      4       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              9.49  Christy Cunningham  3560325168603410  
1              4.45      Douglas Tucker  4478071379779230  
2              8.00      Travis Walters  6011812112971322  
3             12.84    Nathaniel Harris  4676137647685994  
4              7.15        Tonya Carter  4832732618637221

In [544]:
df_tips.iloc[0]

Payment ID                     Sun2959
index                                0
total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  9.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: 0, dtype: object

In [547]:
df_tips = df_tips.set_index('Payment ID')

In [548]:
df_tips.loc['Sun2959']

index                                0
total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  9.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [549]:
df_tips.iloc[0]

index                                0
total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  9.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [551]:
df_tips.iloc[0:4]

index  total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                              
Sun2959         0       16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608         1       10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458         2       21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260         3       23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 9.49  Christy Cunningham  3560325168603410  
Sun4608                 4.45      Douglas Tucker  4478071379779230  
Sun4458                 8.00      Travis Walters  6011812112971322  
Sun5260                12.84    Nathaniel Harris  4676137647685994

In [553]:
df_tips.loc[['Sun2959', 'Sun5260']]

index  total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                              
Sun2959         0       16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260         3       23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 9.49  Christy Cunningham  3560325168603410  
Sun5260                12.84    Nathaniel Harris  4676137647685994

In [556]:
df_tips = df_tips.drop('Sun2959')

In [557]:
df_tips

index  total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                               
Sun4608         1       10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458         2       21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260         3       23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251         4       24.59  3.61  Female     No   Sun  Dinner     4   
Sun9679         5       25.29  4.71    Male     No   Sun  Dinner     4   
...           ...         ...   ...     ...    ...   ...     ...   ...   
Sat2657       239       29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766       240       27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880       241       22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17         242       17.82  1.75    Male     No   Sat  Dinner     2   
Thur672       243       18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun4608                 4.45    Douglas Tucker  4478071379779230  
Sun4458                 8.00    Travis Walters  6011812112971322  
Sun5260                12.84  Nathaniel Harris  4676137647685994  
Sun2251                 7.15      Tonya Carter  4832732618637221  
Sun9679                 7.32        Erik Smith   213140353657882  
...                      ...               ...               ...  
Sat2657                10.68     Michael Avila  5296068606052842  
Sat1766                14.59    Monica Sanders  3506806155565404  
Sat3880                12.34        Keith Wong  6011891618747196  
Sat17                   9.91      Dennis Dixon     4375220550950  
Thur672                10.39   Michelle Hardin  3511451626698139  

[243 rows x 11 columns]

In [561]:
one_row = df_tips.iloc[0]

In [562]:
one_row

index                              1
total_bill                     10.34
tip                             1.66
sex                             Male
smoker                            No
day                              Sun
time                          Dinner
size                               3
price_per_person                4.45
Payer Name            Douglas Tucker
CC Number           4478071379779230
Name: Sun4608, dtype: object

In [573]:
pd.concat([df_tips, pd.DataFrame([one_row])], axis=0)

index  total_bill   tip     sex smoker   day    time  size  \
Sun4608      1       10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458      2       21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260      3       23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251      4       24.59  3.61  Female     No   Sun  Dinner     4   
Sun9679      5       25.29  4.71    Male     No   Sun  Dinner     4   
...        ...         ...   ...     ...    ...   ...     ...   ...   
Sat17      242       17.82  1.75    Male     No   Sat  Dinner     2   
Thur672    243       18.78  3.00  Female     No  Thur  Dinner     2   
Sun4608      1       10.34  1.66    Male     No   Sun  Dinner     3   
Sun4608      1       10.34  1.66    Male     No   Sun  Dinner     3   
Sun4608      1       10.34  1.66    Male     No   Sun  Dinner     3   

         price_per_person        Payer Name         CC Number  
Sun4608              4.45    Douglas Tucker  4478071379779230  
Sun4458              8.00    Travis Walters  6011812112971322  
Sun5260             12.84  Nathaniel Harris  4676137647685994  
Sun2251              7.15      Tonya Carter  4832732618637221  
Sun9679              7.32        Erik Smith   213140353657882  
...                   ...               ...               ...  
Sat17                9.91      Dennis Dixon     4375220550950  
Thur672             10.39   Michelle Hardin  3511451626698139  
Sun4608              4.45    Douglas Tucker  4478071379779230  
Sun4608              4.45    Douglas Tucker  4478071379779230  
Sun4608              4.45    Douglas Tucker  4478071379779230  

[246 rows x 11 columns]

In [574]:
df_tips

index  total_bill   tip     sex smoker   day    time  size  \
Sun4608      1       10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458      2       21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260      3       23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251      4       24.59  3.61  Female     No   Sun  Dinner     4   
Sun9679      5       25.29  4.71    Male     No   Sun  Dinner     4   
...        ...         ...   ...     ...    ...   ...     ...   ...   
Sat3880    241       22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17      242       17.82  1.75    Male     No   Sat  Dinner     2   
Thur672    243       18.78  3.00  Female     No  Thur  Dinner     2   
Sun4608      1       10.34  1.66    Male     No   Sun  Dinner     3   
Sun4608      1       10.34  1.66    Male     No   Sun  Dinner     3   

         price_per_person        Payer Name         CC Number  
Sun4608              4.45    Douglas Tucker  4478071379779230  
Sun4458              8.00    Travis Walters  6011812112971322  
Sun5260             12.84  Nathaniel Harris  4676137647685994  
Sun2251              7.15      Tonya Carter  4832732618637221  
Sun9679              7.32        Erik Smith   213140353657882  
...                   ...               ...               ...  
Sat3880             12.34        Keith Wong  6011891618747196  
Sat17                9.91      Dennis Dixon     4375220550950  
Thur672             10.39   Michelle Hardin  3511451626698139  
Sun4608              4.45    Douglas Tucker  4478071379779230  
Sun4608              4.45    Douglas Tucker  4478071379779230  

[245 rows x 11 columns]